papis/commands/update.py

```diff
@@ -39,25 +107,237 @@
+# TODO: should we just move to having this stored as dict?
+#       - could then also potentially be used for e.g. `whoosh-scheme-prototype`
+#       - or move this function into config (as a specialised getter)?
+KEY_TYPES = papis.commands.doctor.get_key_type_check_keys()
+
+
+def try_formatting_str(
+    key: str, value: str, document: papis.document.DocumentLike
+) -> str:
+    """
+    Tries to format the input string with the currently configured formatter.
+
+    :returns: A string derived from the input string by formatting or an
+        unchanged string.
+    """
+    try:
+        value = papis.format.format(
+            value, document
+        )
+    except papis.format.FormatFailedError:
+        logger.debug("Value '%s' of key '%s' isn't a papis format string.", value, key)
+    return value
+
+
+def try_parsing_str(key: str, value: str) -> str:
+    """
+    Tries to parse the input string as a python expression.
+
+    :returns: The parsed input string if string is a python expression,
+        otherwise an unchanged string.
+    """
+    try:
+        # TODO: elsewhere in the codeabase `eval()` seems to be used
+        # but this seems saver? Something for another PR?
+        value = ast.literal_eval(value)
+    except (SyntaxError, ValueError):
+        logger.debug("Value '%s' of key '%s' is not a python expression.", value, key)
+    return value
+
+
+def is_value_admissible(key: str, value: Any) -> bool:
+    """
+    Checks if the type of the input `value` is what is expected for values of `key`.
+
+    :returns: True if the value is admissible, False if it isn't.
+    """
+    success = True
+    if key in KEY_TYPES and KEY_TYPES[key] != type(value):
+        logger.error(
+            "We got the value '%s' which is of type '%s' but keys '%s' require \
+            type '%s'. No changes have been applied. Aborting...",
+            value,
+            type(value).__name__,
+            key,
+            KEY_TYPES[key].__name__,
+        )
+        success = False
+    return success
+
+
+def run_set(
+    document: papis.document.DocumentLike, to_set: List[Tuple[str, str]]
+) -> Tuple[dict[str, Any], bool]:
+    """
+    Processes a list of `to_set` tuples and applies the resulting changes to the
+    input document.
+
+    :returns: A tuple of the updated document and a boolean indicating whether
+        the update was successful.
+    """
+    processed_tuples: Dict[str, Any] = {}
+    success = True
+    for key, value in to_set:
+        processed_value: Any = try_parsing_str(key, value)
+        print(value)
+        # TODO: I'm aborting if an error is found rather than processing whatever is
+        #       correct. I noticed this is different from some other places in papis
+        #       (e.g. `doctor`). But I think it's better to as obvious as possible when
+        #       stuff is wrong. Let me know if I should change it (same applies to other
+        #       functions).
```

Comment:
Yeah, I think most commands modify what they can and issue warnings for any errors. I'm not sure what's better here..

Reply:
Yeah, I noticed that too. I think it would be better (in general) not to modify anything if there's an error. A reason why this is particularly problematic for the `papis update` command is that re-running the command after fixing the error can lead to undesired results. So rather than just fixing the command you ran, you need to fix it and then remove all the things you are sure ran correctly...